### PR TITLE
Global featured APIs in Branding setting

### LINKED ISSUE #2786: Mark APIs featured on global level
Similar functionality as in issue #2311.

# Goal
* Administrators can select certain APIs to be "Featured APIs" on APInf Catalog page.
* Maximum amount of Featured APIs is 4.
* The Featured APIs 
  * are displayed at the top of Catalog page as their own section.
  * are displayed on Front page instead of Latest APIs (current implementation). (This will a separate task later.)
* If the APIs are sorted by name Featured APIs (in their own section) always precede any other APIs.  Sorting by name means Featured APIs sorted by name followed by all other APIs sorted by name. Likewise for sorting by time added.
* There is a small visual indication of featured API on the API card, which can be toggled on and off. This is visible for admins only.

Refer to the Featured APIs implementation for Organization profile.

# Wireframe
1. Under Branding menu, a new dropdown named "**Featured APIs**" appears in between Site Slogan and Colors controllers. The menu has the following default text "**Select upto 8 APIs to be featured on the Homepage.**"
featuredapihomepage1

2. On clicking the dropdown, site admin can see the list of APIs available in the platform. The API names appear with checkboxes.
featuredapihomepage2

3. On checking an API name, it appears under the dropmenu as [Bootstrap tag](https://bootstrap-tagsinput.github.io/bootstrap-tagsinput/examples/). The dropmenu should remain open on clicking checkboxes as long as user checks 8 boxes. If user needs to close the menu before 8 APIs are checked., s/he should click manually.
featuredapihomepage3

4. When user tries to select more tham 8 APIs, alert text appears with CSS alert class.
featuredapihomepage4

### PR DESCRIPTION
**Pull request by marla-singer**
- Add autoform-select2 package
- Update schema of Branding collection (additional field "featuredApis")
- Display field with multiple select
- Display featured APIs on home page if they exist otherwise display Latest APIs
- New publish functions (apisForBranding & homePageFeaturedApis)
- New template homePageApis & new folder homePageApis under /apis

The main PR is #2906
Current is rebase version
Closes #2906 
Closes #2786

### CODE REVIEW

File: apinf_packages/core/lib/i18n/en.i18n.json [hunk 1]
```diff
@@ -197,6 +197,7 @@
   "branding_panel_title": "Project Branding",
   "branding_projectLogoTitle_logoTitle": "Site Logo",
   "branding_projectCoverPhotoTitle_coverPhotoTitle": "Cover Image",
+  "branding_projectFeaturedApisMessage_featuredApiMessage": "Select upto 8 APIs to be featured in the Homepage",
```
Comment (mauriciovieira) [inline apinf_packages/core/lib/i18n/en.i18n.json]: _select [up to](https://www.usingenglish.com/forum/threads/137090-quot-upto-quot-or-quot-up-to-quot) 8 APIs_

**Comment by bajiat** [comment]
@brylie Could you review this one?

**Comment by brylie** [comment]
Jep.

File: apinf_packages/core/lib/i18n/en.i18n.json (hunk 1 above)
Comment (marla-singer) [reply apinf_packages/core/lib/i18n/en.i18n.json]: @mauriciovieira Fixed

**Comment by brylie** [comment]
I am getting an error related to `aldeed:autoform-select2`:

```
TypeError: template.$(...).select2 is not a function
    at Blaze.TemplateInstance.Template.afSelect2_bootstrap3.rendered
```

**Comment by marla-singer** [comment]
@saransh-dev Do you have any error in browser console? My browser console is clear and couldn't reproduce this error

**Comment by saransh-dev** [comment]
@marla-singer, Same here, my browser console is also clear. No error there.

**Comment by brylie** [comment]
Rather than pinning the `autoform-select2` package to the 1.x branch, which resolves dependency conflicts, we can just solve this without the `autoform-select2` package. I proposed this to @saransh-dev a couple of days ago in chat:

```
<!-- in /apinf_packages/branding/client/branding.html:44 -->
{{> afFieldInput name="featuredApis" multiple=true options=featuredApisOptions }}
```

```js
// in /apinf_packages/branding/client/branding.js
// add the following template helper
...
featuredApisOptions () {
  // Get featuredApis or an empty array on default
  const featuredApis = Branding.findOne().featuredApis || [];

  const featuredApisOptions = Apis.find().map((api) => {
    return {
      text: api.name,
      id: api._id,
        // If featured APIs contains current one then mark it as "selected"
        selected: featuredApis.indexOf(api._id) > -1,
      };
    })

  return featuredApisOptions;
},
```

```js
// in /apinf_packages/branding/client/autoform.js
Template.autoForm.onRendered(function () {
  // Get ID of current AutoForm
  const formId = this.data.id;

  if (formId === 'brandingEdit') {
    // Get localized text for featured APIs field
   const message = TAPi18n.__('branding_projectFeaturedApisMessage_featuredApiMessage');
    
    // Render select2 widget on Featured APIs field
    $('[name=featuredApis]').select2({
      placeholder: message,
    });

  }
});
```

**Comment by brylie** [comment]
Admittedly, using the `aldeed:autoform-select2` package would be ideal, since it is the most declarative solution. However, pinning the version at 1.x to solve the version constraints does not seem ideal. Hence, the proposal to remove `aldeed:autoform-select2` and use an imperative `autoform.onRendered` callback.

**Comment by saransh-dev** [comment]
Hi @brylie, i have used above solution as you provided to me but there was problem, after submit it displayed default `select-multiple`. Please review below gif:-
sep-28-2017 17-26-39

**Comment by brylie** [comment]
Did you add the `autoform.onRendered` callback? Also, make sure the AutoForm does not render until template subscriptions are ready.

**Comment by brylie** [comment]
@saransh-dev, as per our chat, I suggested to use the following `afFieldInput`:

```
{{> afFieldInput name="featuredApis" multiple=true options=featuredApisOptions }}
``

**Comment by saransh-dev** [comment]
yeah, at that time i used `{{> afFieldInput name="featuredApis" multiple=true options=featuredApisOptions }}` i chenged it  after the final result for label. Because with the use of `afFieldInput`, label could not display. So i changed it to `afQuickField`.

**Comment by brylie** [comment]
OK, make sure the `afQuickField` is using a consistent field name, otherwise jQuery might not be able to find the field:

```js
// Render select2 widget on Featured APIs field
    $('[name=featuredApis]').select2({
      placeholder: message,
    });
```

**Comment by brylie** [comment]
> with the use of afFieldInput, label could not display.

With `afFieldInputs` we just manually handle the label text:

https://github.com/apinf/platform/blob/38bc28a07a974cc0d878d53559acc1ca6cf879a9/apinf_packages/monitoring/client/monitoring.html#L32-L38

**Comment by saransh-dev** [comment]
Yes, it is using same field name as "featuredApis"
 ```
<select name="featuredApis" multiple="" id="skgKWmZucdSn2kD7W" data-schema-key="featuredApis" autocomplete="off" class="form-control select2-hidden-accessible" tabindex="-1" aria-hidden="true">
</select>
```

**Comment by saransh-dev** [comment]
Yes, we can handle it manually for `afFieldInput`.

**Comment by marla-singer** [comment]
@brylie After updating featured APIs list `Template.autoForm.onRendered(function () {` didn't trigger. It runs once after rendering page and all. Because of solution doesn't work properly

**Comment by marla-singer** [comment]
@maur I fixed type changes

**Comment by brylie** [comment]
Ok, sorry for not catching that before. I guess the autoform-select2 1.x is the best solution for now.

**Comment by bajiat** [comment]
Is there a need  for a database migration for this issue? If yes, has it been implemented?

**Comment by marla-singer** [comment]
> Is there a need for a database migration for this issue? If yes, has it been implemented?

@bajiat No, it doesn't need migration. The new field isn't required and doesn't have a default value. When any value will be assigned then it will appear

File: apinf_packages/branding/client/branding.js [hunk 2]
```diff
@@ -21,4 +38,23 @@ Template.branding.helpers({
   brandingCollection () {
     return Branding;
   },
+  s2Opts () {
```
[the remark this answers is not in the document]
Comment (mauriciovieira) [reply apinf_packages/branding/client/branding.js]: So, my final suggestion is to rename to `homepageApisSelect2Options`

**Comment by bajiat** [comment]
@saransh-dev Have you worked on the requested changes? Also please remember to rebase.

**Comment by marla-singer** [comment]
@bajiat I will fix the changes

**Comment by marla-singer** [comment]
@mauriciovieira Done

**Comment by bajiat** [comment]
@mauriciovieira Can you review and merge, if ok?

**Comment by bajiat** [comment]
@matleppa Can you review the PR?

**Comment by matleppa** [comment]
OK

**Comment by matleppa** [comment]
I keep getting error on log when opening the Branding page.

TypeError: template.$(...).select2 is not a function
    at Blaze.TemplateInstance.Template.afSelect2_bootstrap3.rendered (http://localhost:3000/packages/aldeed_autoform-select2.js?hash=e60cfa8faab90aa51ff8594b34a7a199eccc00d4:272:24)
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3398:22
    at Function.Template._withTemplateInstanceFunc (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3744:12)
    at fireCallbacks (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3394:12)
    at Blaze.View.<anonymous> (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3487:5)
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:1845:14
    at Object.Blaze._withCurrentView (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:2271:12)
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:1844:15
    at Tracker._runFlush (http://localhost:3000/packages/tracker.js?hash=997515fa2d5b0530ba07741da556c4b36963ef3b:548:11)
    at e.exports (http://localhost:3000/swagger/bundle.js:1:8056)
    at w.(anonymous function) (http://localhost:3000/swagger/bundle.js:105:7780)
    at Number.S (http://localhost:3000/swagger/bundle.js:105:7638)
    at MessagePort.E (http://localhost:3000/swagger/bundle.js:105:7660)

**Comment by marla-singer** [comment]
@matleppa Install pacakges and try again
> npm i
> meteor

**Comment by matleppa** [comment]
I've been switching the branch on and off, and using meteor npm install command, however, no help.
Strange.


C:\Users\matleppa\api\apinf\platform>meteor list
...
aldeed:autoform-select2                       1.0.5* Custom select2 input type for AutoForm
aldeed:collection2                            2.10.0  Automatic validation of insert and update operations on th...



### package.json
  "name": "apinf",
  "version": "0.50.0",
  "description": "API management portal and proxy.",
  "main": "index.js",
  "directories": {
    "doc": "docs"
  },
  "dependencies": {
    "JSONStream": "^1.3.1",
    "babel-runtime": "^6.22.0",
    "bcrypt": "^0.8.7",
    "chart.js": "^2.6.0",
    "clipboard": "^1.5.16",
    "crossfilter": "^1.3.12",
    "d3": "^3.5.17",
    "datatables": "^1.10.12",
    "dc": "^2.1.3",
    "elasticsearch": "^11.0.1",
    "github": "^1.4.0",
    "got": "^6.7.1",
    "intro.js": "^2.4.0",
    "jquery": "^2.2.4",
    "jquery-bbq": "^1.0.0",
    "jquery.rateit": "^1.0.23",
    "js-yaml": "^3.8.1",
    "json2csv": "^3.11.1",
    "locale-compare-polyfill": "0.0.2",
    "lodash": "^4.17.4",
    "meteor-node-stubs": "^0.2.5",
    "moment": "^2.17.1",
    "raven": "^2.1.0",
    "raven-js": "^3.16.0",
    **"select2": "^4.0.3",**
    "select2-bootstrap-theme": "0.1.0-beta.10",
    "simple-statistics": "^2.2.0",
...

**Comment by marla-singer** [comment]
@matleppa Look at the node_modules folder. Does it contain "select2" sub-folder?

**Comment by marla-singer** [comment]
@matleppa The same stuff was with @brylie. Hm. I have no idea about why it happens

**Comment by matleppa** [comment]
> @matleppa Look at the node_modules folder. Does it contain "select2" sub-folder?

Yes. it was there.
Now I re-installed the local environment, but still the exception occurs.

Exception from Tracker afterFlush function:
meteor.js:942 TypeError: template.$(...).select2 is not a function
    at Blaze.TemplateInstance.Template.afSelect2_bootstrap3.rendered (http://localhost:3000/packages/aldeed_autoform-select2.js?hash=e60cfa8faab90aa51ff8594b34a7a199eccc00d4:272:24)
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3398:22
    at Function.Template._withTemplateInstanceFunc (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3744:12)
    at fireCallbacks (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3394:12)
    at Blaze.View.<anonymous> (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3487:5)
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:1845:14
    at Object.Blaze._withCurrentView (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:2271:12)
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:1844:15
    at Tracker._runFlush (http://localhost:3000/packages/tracker.js?hash=997515fa2d5b0530ba07741da556c4b36963ef3b:548:11)
    at e.exports (http://localhost:3000/swagger/bundle.js:1:8056)
    at w.(anonymous function) (http://localhost:3000/swagger/bundle.js:105:7780)
    at Number.S (http://localhost:3000/swagger/bundle.js:105:7638)
    at MessagePort.E (http://localhost:3000/swagger/bundle.js:105:7660)

**Comment by marla-singer** [comment]
@matleppa Fixed it. Please check again

**Comment by bajiat** [comment]
In my opinion the view should contain More APIs button or similar CTA to go to Catalog. 

@55 Any opinions?

**Comment by 55** [comment]
@bajiat seems legit, let's go with `More APIs` button or similar CTA to go to Catalog.

**Comment by marla-singer** [comment]
@saransh-dev Will you have time to take the button and should I finish it?

**Comment by saransh-dev** [comment]
@marla-singer  @matleppa Admin can select only upto 8 featured apis for home page. So what is the use of `More APIs` button ?

Please advise.

**Comment by marla-singer** [comment]
@saransh-dev You need to add button "More APIs" on the home page in the block "Featured APIs" if this block exists
joxi_screenshot_1508403541728

**Comment by saransh-dev** [comment]
@marla-singer So i need to provide route of api catalog page on this button. Right ?

**Comment by marla-singer** [comment]
Yes. The block "Latest APIs" already has this button, just duplicate it

**Comment by saransh-dev** [comment]
ok Can i do it tomorrow. Because today i am on leave. Today we have national festival [Diwali](https://en.wikipedia.org/wiki/Diwali).

**Comment by marla-singer** [comment]
@saransh-dev No problem

**Comment by matleppa** [comment]
Instead of duplicating the "Latest APIs" button, why not just move the code, which generates the button, to upper level inside the template. 
When the button program code is located after the featured APIs / latest APIs selection, it can be used in both cases.

**Comment by bajiat** [comment]
@mauriciovieira Can you approve the changes you had requested?

**Comment by marla-singer** [comment]
@saransh-dev It is strange in the commit history. I will squash it all 4 commits together
joxi_screenshot_1508502824597

**Comment by marla-singer** [comment]
@saransh-dev Umh. Looks like you made `git pull` and it created a message about merge and you squashed it as well. 
In the future, try to run `git pull --rebase` that allows to pull changes w/out a commit about merge.
I will add your changes by my hand because it doesn't cherry-pick your commit :confused:

**Comment by saransh-dev** [comment]
@marla-singer, I just did following steps:-
1. checkout to `develop` and took pull from `develop`.
2. checkout to `feature/global-featured-apis` and took pull from `feature/global-featured-apis`.
3. worked on `feature/global-featured-apis` branch.
4. `git add .` to add changes then commit a message after that `git pull origin feature/global-featured-apis and finally `git push origin feature/global-featured-apis`

**Comment by marla-singer** [comment]
@saransh-dev Did you do `git rebase develop -i`?

**Comment by saransh-dev** [comment]
No, I didn't do

**Comment by marla-singer** [comment]
Then it is a magic why you commit contains this code
joxi_screenshot_1508503920517

**Comment by marla-singer** [comment]
anyway, at least it works.
@matleppa Please approve it as well

**Comment by saransh-dev** [comment]
I just got conflicts while pull from `feature/global-featured-apis`. In that I found `natestreauser:select2` package in packages file

terminal

**Pull request merged**